pkg/neonvm/controllers/vm_controller_cpu_scaling.go

```diff
@@ -0,0 +1,119 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	corev1 "k8s.io/api/core/v1"
+
+	vmv1 "github.com/neondatabase/autoscaling/neonvm/apis/neonvm/v1"
+	"github.com/neondatabase/autoscaling/pkg/api"
+)
+
+// handleCPUScaling is extracted from doReconcile and encapsulates the logic to handle CPU scaling.
+// if vm scaling mode is set to CpuSysfsState, the scaling is delegated to neonvm-daemon
+// otherwise the scaling is first done by scaling amount of cores in the VM using QMP and then by updating the cgroup
+func (r *VMReconciler) handleCPUScaling(ctx context.Context, vm *vmv1.VirtualMachine, vmRunner *corev1.Pod) (bool, error) {
+	log := log.FromContext(ctx)
+	useCpuSysfsStateScaling := *vm.Spec.CpuScalingMode == vmv1.CpuScalingModeSysfs
+
+	var scaled bool
+	var err error
+	if !useCpuSysfsStateScaling {
+		scaled, err = r.handleCPUScalingQMP(ctx, vm, vmRunner)
+	} else {
+		scaled, err = r.handleCPUScalingSysfs(ctx, vm, vmRunner)
+	}
+
+	if err != nil {
+		log.Error(err, "Failed to scale CPU", "VirtualMachine", vm.Name, "CpuScalingMode", vm.Spec.CpuScalingMode)
+		return false, err
+	}
+
+	return scaled, nil
+}
+
+// handleCPUScalingQMP handles CPU scaling using QMP, extracted as is from doReconcile
+func (r *VMReconciler) handleCPUScalingQMP(ctx context.Context, vm *vmv1.VirtualMachine, vmRunner *corev1.Pod) (bool, error) {
```

Comment:
suggestion: This function is very visually dense. Could you break it up more by adding newlines between sections?

Reply:
tried to improve readability here.